tools/chapel-py/README.md

```diff
@@ -0,0 +1,458 @@
+
+positional arguments:
+  filenames
+
+options:
+  -h, --help       show this help message and exit
+  --suffix SUFFIX
+  --in-place
+```
+
+The 'source of changes' is a Python iterator that should accept two arguments:
+`rc` (for 'replacement context') and `root` (for the AST node at which the
+traversal begins). A very simple (albeit completely pointless and semantically
+incorrect) example is the following script, which replaces all occurrences of
+the number `42` with `meaningOfLife`:
+
+```Python
+def replace_all_42(rc, root):
+    for (num, _) in each_matching(root, IntLiteral):
+        if num.text() == '42': yield (num, 'meaningOfLife')
+
+run(replace_all_42)
+```
+
+A more practical example that of renaming methods. Note that this is not
+amenable to naive substitution: we don't want to accidentally rename non-methods
+that happen to have the same name, or even locally declared functions inside
+other functions in a record or class. Consider renaming the `enterThis`
+method to `enterContext` (this is a real change that was performed in
+the Chapel language in 1.32, though it was not done automatically with the
+help of this tool). In the following snippet, only the `proc` labeled `(1)`
+should be renamed. This is because `enterThis` is actually a method. On
+the other hand, `(2)`, even though it's declared within `R`'s curly braces,
+is actually a locally defined function, and not a method on `R`. Of course,
+`(3)` is a freestanding function, and thus should also not be renamed.
+
+```Chapel
+record R {
+    proc enterThis() {} // (1)
+    proc f() {
+        proc enterThis() {} // (2)
+    }
+}
+proc enterThis() {} // (3)
+```
+
+This can be implemented using the following iterator:
+
+```Python
+def replace_enter_this(rc, root):
+    for (fn, _) in chapel.each_matching(root, Function):
+        if fn.name() == "enterThis" and fn.is_method():
+            yield (fn, lambda text: text.replace("enterThis", "enterContext", 1))
+```
+
+In this case, instead of yielding a pair of the node-to-replace and the
+replacement text, the iterator produces a `lambda`, which will be fed the
+node's current string representation (`proc enterThis() {...`). Using
+this lambda, we perform a simple substitution, renaming `enterThis` to
+`enterContext` with Python's regular `replace` method.
+
+Finally, the following script was used to automatically insert interfaces into
+records and classes that provided the corresponding functionality. It was
+used to update about 150 files (including Chapel modules and tests). This
+script implements the following features:
+
+* Detecting particular signatures of the `init` method specific to deserialization
+* Always printing the supported interfaces in a specific order
+* Supporting both primary and secondary methods, including a mix of both
+* Adding a new interface list or modifying an existing one
+
+```Python
+def tag_aggregates_with_io_interfaces(rc, root):
```

Comment:
It would work on GitHub, but not in independent / offline clones, which I think would bother more more than drift. That's the reason I included it verbatim.